Number Precision: RecordAgingNotifier fails unmarshaling contracts causing db failure

On the number precision branch, existing contracts from the sales and purchases table fail with:
`[ERROR] [recordAgingNotifier/PerformTask] generateBuyerDisputeTimeoutNotifications failed: unmarshaling contract: json: cannot unmarshal string into Go value of type map[string]json.RawMessage` 

Once this happens, lot of db errors are logged. Some examples:
`[ERROR] [recordAgingNotifier/PerformTask] generateSellerDisputeNotifications failed: cannot start a transaction within a transaction`

This also results in failures to write to the db:
```
[DEBUG] [transaction-listener/processPurchasePayment] Payment for purchase QmbdXw8kMHV1VQ8L88vaTCpM72DcRJxsxixinp1G9uJdsw detected
goroutine 11988 [running]:
runtime/debug.Stack(0xc0009ed180, 0x0, 0x0)
	/src/runtime/debug/stack.go:24 +0x9d
runtime/debug.PrintStack()
	/src/runtime/debug/stack.go:16 +0x22
github.com/OpenBazaar/openbazaar-go/api.(*jsonAPIHandler).ServeHTTP.func1()
	/github.com/OpenBazaar/openbazaar-go/api/jsonapi.go:165 +0x145
panic(0x21478c0, 0x37a34a0)
	/src/runtime/panic.go:522 +0x1b5
sync.(*RWMutex).RLock(...)
	/src/sync/rwmutex.go:48
database/sql.(*Tx).grabConn(0x0, 0x28ecc80, 0xc0000c6010, 0x28ecc80, 0xc0000c6010, 0x0, 0xc000c8f201)
	/src/database/sql/sql.go:1958 +0x58
database/sql.(*Tx).PrepareContext(0x0, 0x28ecc80, 0xc0000c6010, 0x23ee3f8, 0x6a, 0x28b0360, 0xc001618e40, 0x40)
	/src/database/sql/sql.go:2053 +0x4c
database/sql.(*Tx).Prepare(...)
	/src/database/sql/sql.go:2075
github.com/OpenBazaar/openbazaar-go/repo/db.(*TxMetadataDB).Put(0xc000508690, 0xc0012a5dc0, 0x40, 0xc002323a10, 0x2a, 0xc000cda81a, 0x2, 0xc002322090, 0x2e, 0xc0023223f0, ...)
	/github.com/OpenBazaar/openbazaar-go/repo/db/txmetadata.go:22 +0xfc
github.com/OpenBazaar/openbazaar-go/core.(*OpenBazaarNode).Spend(0xc0011c6e00, 0xc0009e2c80, 0xc0009e2c80, 0x0, 0x0)
	/github.com/OpenBazaar/openbazaar-go/core/spend.go:118 +0x80a
github.com/OpenBazaar/openbazaar-go/api.(*jsonAPIHandler).POSTSpendCoinsForOrder(0xc0011c6f00, 0x28e5300, 0xc00220a2a0, 0xc00168c000)
	/github.com/OpenBazaar/openbazaar-go/api/jsonapi.go:765 +0x130
```

--- a/core/record_aging_notifier.go
+++ b/core/record_aging_notifier.go
@@ -179,7 +179,11 @@ func (notifier *recordAgingNotifier) generateSellerDisputeNotifications() (*noti
 	if err = notifier.datastore.Sales().UpdateSalesLastDisputeTimeoutNotifiedAt(updatedSales); err != nil {
 		return nil, fmt.Errorf("update sales disputeTimeoutNotifiedAt: %s", err.Error())
 	}
-	return &notifierResult{len(notificationsToAdd), len(updatedSales), "sales"}, nil
+	return &notifierResult{
+		notificationsMade: len(notificationsToAdd),
+		recordsUpdated:    len(updatedSales),
+		subject:           "sales",
+	}, nil
 }
 
 func (notifier *recordAgingNotifier) generateBuyerDisputeTimeoutNotifications() (*notifierResult, error) {
@@ -258,7 +262,11 @@ func (notifier *recordAgingNotifier) generateBuyerDisputeTimeoutNotifications() 
 	if err = notifier.datastore.Purchases().UpdatePurchasesLastDisputeTimeoutNotifiedAt(updatedPurchases); err != nil {
 		return nil, fmt.Errorf("updating lastDisputeTimeoutNotifiedAt on purchases: %s", err.Error())
 	}
-	return &notifierResult{len(notificationsToAdd), len(updatedPurchases), "purchaseTimeout"}, nil
+	return &notifierResult{
+		notificationsMade: len(notificationsToAdd),
+		recordsUpdated:    len(updatedPurchases),
+		subject:           "purchaseTimeout",
+	}, nil
 }
 
 func (notifier *recordAgingNotifier) generateBuyerDisputeExpiryNotifications() (*notifierResult, error) {
@@ -333,7 +341,11 @@ func (notifier *recordAgingNotifier) generateBuyerDisputeExpiryNotifications() (
 	if err = notifier.datastore.Purchases().UpdatePurchasesLastDisputeExpiryNotifiedAt(updatedPurchases); err != nil {
 		return nil, fmt.Errorf("updating lastDisputeExpiryNotifiedAt on purchases: %s", err.Error())
 	}
-	return &notifierResult{len(notificationsToAdd), len(updatedPurchases), "purchaseExpire"}, nil
+	return &notifierResult{
+		notificationsMade: len(notificationsToAdd),
+		recordsUpdated:    len(updatedPurchases),
+		subject:           "purchaseExpire",
+	}, nil
 }
 
 func (notifier *recordAgingNotifier) generateModeratorDisputeExpiryNotifications() (*notifierResult, error) {
@@ -412,5 +424,9 @@ func (notifier *recordAgingNotifier) generateModeratorDisputeExpiryNotifications
 	if err = notifier.datastore.Cases().UpdateDisputesLastDisputeExpiryNotifiedAt(updatedDisputes); err != nil {
 		return nil, fmt.Errorf("updating lastDisputeExpiryNotifiedAt on disputes: %s", err.Error())
 	}
-	return &notifierResult{len(notificationsToAdd), len(updatedDisputes), "dispute"}, nil
+	return &notifierResult{
+		notificationsMade: len(notificationsToAdd),
+		recordsUpdated:    len(updatedDisputes),
+		subject:           "dispute",
+	}, nil
 }

--- a/core/utils.go
+++ b/core/utils.go
@@ -104,19 +104,9 @@ func (n *OpenBazaarNode) BuildTransactionRecords(contract *pb.RicardianContract,
 	return paymentRecords, refundRecord, nil
 }
 
-// LookupCurrency looks up the CurrencyDefinition, first by crypto for the current network
-// (mainnet or testnet) and then by fiat code
+// LookupCurrency looks up the CurrencyDefinition from available currencies
 func (n *OpenBazaarNode) LookupCurrency(currencyCode string) (repo.CurrencyDefinition, error) {
-	if n.TestnetEnable || n.RegressionTestEnable {
-		if def, err := repo.TestnetCurrencies().Lookup(currencyCode); err == nil {
-			return def, nil
-		}
-	} else {
-		if def, err := repo.MainnetCurrencies().Lookup(currencyCode); err == nil {
-			return def, nil
-		}
-	}
-	return repo.FiatCurrencies().Lookup(currencyCode)
+	return repo.AllCurrencies().Lookup(currencyCode)
 }
 
 // exchangeRateCode strips the T off the currency code if we are on testnet or regtest.

--- a/repo/db/cases.go
+++ b/repo/db/cases.go
@@ -31,17 +31,13 @@ func (c *CasesDB) PutRecord(dispute *repo.DisputeCaseRecord) error {
 		buyerOpenedInt = 1
 	}
 
-	tx, err := c.db.Begin()
-	if err != nil {
-		return err
-	}
 	stm := `insert or replace into cases(caseID, state, read, timestamp, buyerOpened, claim, buyerPayoutAddress, vendorPayoutAddress, paymentCoin, coinType) values(?,?,?,?,?,?,?,?,?,?)`
-	stmt, err := tx.Prepare(stm)
+	stmt, err := c.PrepareQuery(stm)
 	if err != nil {
 		return err
 	}
-
 	defer stmt.Close()
+
 	_, err = stmt.Exec(
 		dispute.CaseID,
 		int(dispute.OrderState),
@@ -55,14 +51,10 @@ func (c *CasesDB) PutRecord(dispute *repo.DisputeCaseRecord) error {
 		dispute.CoinType,
 	)
 	if err != nil {
-		rErr := tx.Rollback()
-		if rErr != nil {
-			return fmt.Errorf("case put fail: %s\nand rollback failed: %s\n", err.Error(), rErr.Error())
-		}
-		return err
+		return fmt.Errorf("update dispute case: %s", err.Error())
 	}
 
-	return tx.Commit()
+	return nil
 }
 
 func (c *CasesDB) Put(caseID string, state pb.OrderState, buyerOpened bool, claim string, paymentCoin string, coinType string) error {
@@ -321,21 +313,24 @@ func (c *CasesDB) GetAll(stateFilter []pb.OrderState, searchTerm string, sortByA
 func (c *CasesDB) GetCaseMetadata(caseID string) (buyerContract, vendorContract *pb.RicardianContract, buyerValidationErrors, vendorValidationErrors []string, state pb.OrderState, read bool, timestamp time.Time, buyerOpened bool, claim string, resolution *pb.DisputeResolution, err error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	stmt, err := c.db.Prepare("select buyerContract, vendorContract, buyerValidationErrors, vendorValidationErrors, state, read, timestamp, buyerOpened, claim, disputeResolution from cases where caseID=?")
+	stmt, err := c.PrepareQuery("select buyerContract, vendorContract, buyerValidationErrors, vendorValidationErrors, state, read, timestamp, buyerOpened, claim, disputeResolution from cases where caseID=?")
 	if err != nil {
 		return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, err
 	}
 	defer stmt.Close()
-	var buyerCon []byte
-	var vendorCon []byte
-	var buyerErrors []byte
-	var vendorErrors []byte
-	var stateInt int
-	var readInt *int
-	var ts int
-	var buyerOpenedInt int
-	var disputResolution []byte
-	err = stmt.QueryRow(caseID).Scan(&buyerCon, &vendorCon, &buyerErrors, &vendorErrors, &stateInt, &readInt, &ts, &buyerOpenedInt, &claim, &disputResolution)
+
+	var (
+		buyerCon          []byte
+		vendorCon         []byte
+		buyerErrors       []byte
+		vendorErrors      []byte
+		stateInt          int
+		readInt           *int
+		ts                int
+		buyerOpenedInt    int
+		disputeResolution []byte
+	)
+	err = stmt.QueryRow(caseID).Scan(&buyerCon, &vendorCon, &buyerErrors, &vendorErrors, &stateInt, &readInt, &ts, &buyerOpenedInt, &claim, &disputeResolution)
 	if err != nil {
 		return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, err
 	}
@@ -371,21 +366,21 @@ func (c *CasesDB) GetCaseMetadata(caseID string) (buyerContract, vendorContract 
 	if string(buyerErrors) != "" {
 		err = json.Unmarshal(buyerErrors, &berr)
 		if err != nil {
-			return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, err
+			return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, fmt.Errorf("unmarshal dispute case: %s", err.Error())
 		}
 	}
 	var verr []string
 	if string(vendorErrors) != "" {
 		err = json.Unmarshal(vendorErrors, &verr)
 		if err != nil {
-			return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, err
+			return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, fmt.Errorf("unmarshal dispute vendor errors: %s", err.Error())
 		}
 	}
 	resolution = new(pb.DisputeResolution)
-	if string(disputResolution) != "" {
-		err = jsonpb.UnmarshalString(string(disputResolution), resolution)
+	if string(disputeResolution) != "" {
+		err = jsonpb.UnmarshalString(string(disputeResolution), resolution)
 		if err != nil {
-			return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, err
+			return nil, nil, []string{}, []string{}, pb.OrderState(0), false, time.Time{}, false, "", nil, fmt.Errorf("unmarhsal dispute case resolution: %s", err.Error())
 		}
 	} else {
 		resolution = nil
@@ -411,10 +406,12 @@ func (c *CasesDB) GetByCaseID(caseID string) (*repo.DisputeCaseRecord, error) {
 		vendorOuts       []byte
 	)
 
-	stmt, err := c.db.Prepare("select buyerContract, vendorContract, buyerPayoutAddress, vendorPayoutAddress, buyerOutpoints, vendorOutpoints, state, buyerOpened, timestamp, paymentCoin from cases where caseID=?")
+	stmt, err := c.PrepareQuery("select buyerContract, vendorContract, buyerPayoutAddress, vendorPayoutAddress, buyerOutpoints, vendorOutpoints, state, buyerOpened, timestamp, paymentCoin from cases where caseID=?")
 	if err != nil {
 		return nil, err
 	}
+	defer stmt.Close()
+
 	err = stmt.QueryRow(caseID).Scan(&buyerCon, &vendorCon, &buyerAddr, &vendorAddr, &buyerOuts, &vendorOuts, &stateInt, &isBuyerInitiated, &createdAt, &paymentCoin)
 	if err != nil {
 		return nil, err
@@ -566,18 +563,24 @@ func (c *CasesDB) UpdateDisputesLastDisputeExpiryNotifiedAt(disputeCases []*repo
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
-	tx, err := c.db.Begin()
+	tx, err := c.BeginTransaction()
 	if err != nil {
 		return fmt.Errorf("begin update disputes transaction: %s", err.Error())
 	}
 	for _, d := range disputeCases {
 		_, err = tx.Exec("update cases set lastDisputeExpiryNotifiedAt = ? where caseID = ?", int(d.LastDisputeExpiryNotifiedAt.Unix()), d.CaseID)
 		if err != nil {
+			if rErr := tx.Rollback(); rErr != nil {
+				return fmt.Errorf("update dispute case: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+			}
 			return fmt.Errorf("update dispute case: %s", err.Error())
 		}
 	}
 	if err = tx.Commit(); err != nil {
-		return fmt.Errorf("commit update disputes transaction: %s", err.Error())
+		if rErr := tx.Rollback(); rErr != nil {
+			return fmt.Errorf("commit dispute case: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+		}
+		return fmt.Errorf("commit disputes case: %s", err.Error())
 	}
 
 	return nil

--- a/repo/db/chat.go
+++ b/repo/db/chat.go
@@ -28,15 +28,13 @@ func (c *ChatDB) Put(messageId string, peerId string, subject string, message st
 		timestamp = time.Now()
 	}
 
-	tx, err := c.db.Begin()
-	if err != nil {
-		return err
-	}
 	stm := `insert into chat(messageID, peerID, subject, message, read, timestamp, outgoing) values(?,?,?,?,?,?,?)`
-	stmt, err := tx.Prepare(stm)
+	stmt, err := c.PrepareQuery(stm)
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare chat sql: %s", err.Error())
 	}
+	defer stmt.Close()
+
 	readInt := 0
 	if read {
 		readInt = 1
@@ -47,7 +45,6 @@ func (c *ChatDB) Put(messageId string, peerId string, subject string, message st
 		outgoingInt = 1
 	}
 
-	defer stmt.Close()
 	_, err = stmt.Exec(
 		messageId,
 		peerId,
@@ -58,15 +55,7 @@ func (c *ChatDB) Put(messageId string, peerId string, subject string, message st
 		outgoingInt,
 	)
 	if err != nil {
-		err = tx.Rollback()
-		if err != nil {
-			log.Error(err)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit chat: %s", err.Error())
 	}
 	return nil
 }
@@ -205,15 +194,8 @@ func (c *ChatDB) MarkAsRead(peerID string, subject string, outgoing bool, messag
 		updateArgs = append(updateArgs, messageId)
 	}
 
-	tx, err := c.db.Begin()
+	result, err := c.db.Exec("update chat set read=1 where subject=? and outgoing=?"+peerStm+messageStm, updateArgs...)
 	if err != nil {
-		return "", false, fmt.Errorf("begin tx mark chat as read: %s", err)
-	}
-	result, err := tx.Exec("update chat set read=1 where subject=? and outgoing=?"+peerStm+messageStm, updateArgs...)
-	if err != nil {
-		if rErr := tx.Rollback(); rErr != nil {
-			return "", false, fmt.Errorf("mark chat as read: %s (rollback: %s)", err, rErr)
-		}
 		return "", false, fmt.Errorf("mark chat as read: %s", err)
 	}
 	if count, err := result.RowsAffected(); err != nil {
@@ -223,14 +205,11 @@ func (c *ChatDB) MarkAsRead(peerID string, subject string, outgoing bool, messag
 			updated = true
 		}
 	}
-	if err := tx.Commit(); err != nil {
-		return "", false, fmt.Errorf("commit tx mark chat as read: %s", err)
-	}
 
 	// get last message ID
 	stmt2, err := c.db.Prepare("select max(timestamp), messageID from chat where subject=?" + peerStm + " and outgoing=?")
 	if err != nil {
-		return "", updated, err
+		return "", updated, fmt.Errorf("prepare get last message id sql: %s", err.Error())
 	}
 	defer stmt2.Close()
 	var (
@@ -243,7 +222,7 @@ func (c *ChatDB) MarkAsRead(peerID string, subject string, outgoing bool, messag
 		err = stmt2.QueryRow(subject, outgoingInt).Scan(&timestamp, &msgId)
 	}
 	if err != nil {
-		return "", updated, err
+		return "", updated, fmt.Errorf("query get last message id: %s", err.Error())
 	}
 	return msgId.String, updated, nil
 }

--- a/repo/db/coupons.go
+++ b/repo/db/coupons.go
@@ -2,6 +2,7 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
 	"sync"
 
 	"github.com/OpenBazaar/openbazaar-go/repo"
@@ -18,22 +19,24 @@ func NewCouponStore(db *sql.DB, lock *sync.Mutex) repo.CouponStore {
 func (c *CouponDB) Put(coupons []repo.Coupon) error {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	tx, _ := c.db.Begin()
+	tx, _ := c.BeginTransaction()
 	for _, coupon := range coupons {
 		stmt, _ := tx.Prepare("insert or replace into coupons(slug, code, hash) values(?,?,?)")
 		defer stmt.Close()
+
 		_, err := stmt.Exec(coupon.Slug, coupon.Code, coupon.Hash)
 		if err != nil {
-			err = tx.Rollback()
-			if err != nil {
-				log.Error(err)
+			if rErr := tx.Rollback(); rErr != nil {
+				return fmt.Errorf("add coupon: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
 			}
-			return err
+			return fmt.Errorf("add coupon: %s", err.Error())
 		}
 	}
-	err := tx.Commit()
-	if err != nil {
-		log.Error(err)
+	if err := tx.Commit(); err != nil {
+		if rErr := tx.Rollback(); rErr != nil {
+			return fmt.Errorf("commit coupon: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+		}
+		return fmt.Errorf("commit coupon: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/db.go
+++ b/repo/db/db.go
@@ -2,6 +2,7 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
 	"path"
 	"sync"
 	"time"
@@ -65,7 +66,7 @@ func Create(repoPath, password string, testnet bool, coinType wallet.CoinType) (
 
 func NewSQLiteDatastore(db *sql.DB, l *sync.Mutex, coinType wallet.CoinType) *SQLiteDatastore {
 	return &SQLiteDatastore{
-		config:          &ConfigDB{db: db, lock: l},
+		config:          &ConfigDB{modelStore{db: db, lock: l}},
 		followers:       NewFollowerStore(db, l),
 		following:       NewFollowingStore(db, l),
 		offlineMessages: NewOfflineMessageStore(db, l),
@@ -241,8 +242,7 @@ func initDatabaseTables(db *sql.DB, password string) (err error) {
 }
 
 type ConfigDB struct {
-	db   *sql.DB
-	lock *sync.Mutex
+	modelStore
 }
 
 func (c *ConfigDB) Init(mnemonic string, identityKey []byte, password string, creationDate time.Time) error {
@@ -251,42 +251,23 @@ func (c *ConfigDB) Init(mnemonic string, identityKey []byte, password string, cr
 	if err := initDatabaseTables(c.db, password); err != nil {
 		return err
 	}
-	tx, err := c.db.Begin()
-	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("insert into config(key, value) values(?,?)")
+	stmt, err := c.PrepareQuery("insert into config(key, value) values(?,?)")
 	if err != nil {
 		return err
 	}
 	defer stmt.Close()
+
 	_, err = stmt.Exec("mnemonic", mnemonic)
 	if err != nil {
-		err = tx.Rollback()
-		if err != nil {
-			log.Error(err)
-		}
-		return err
+		return fmt.Errorf("set mnemonic: %s", err.Error())
 	}
 	_, err = stmt.Exec("identityKey", identityKey)
 	if err != nil {
-		err = tx.Rollback()
-		if err != nil {
-			log.Error(err)
-		}
-		return err
+		return fmt.Errorf("set identity key: %s", err.Error())
 	}
 	_, err = stmt.Exec("creationDate", creationDate.Format(time.RFC3339))
 	if err != nil {
-		err = tx.Rollback()
-		if err != nil {
-			log.Error(err)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("set creation date: %s", err.Error())
 	}
 	return nil
 }
@@ -294,7 +275,7 @@ func (c *ConfigDB) Init(mnemonic string, identityKey []byte, password string, cr
 func (c *ConfigDB) GetMnemonic() (string, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	stmt, err := c.db.Prepare("select value from config where key=?")
+	stmt, err := c.PrepareQuery("select value from config where key=?")
 	if err != nil {
 		log.Fatal(err)
 	}
@@ -310,7 +291,7 @@ func (c *ConfigDB) GetMnemonic() (string, error) {
 func (c *ConfigDB) GetIdentityKey() ([]byte, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	stmt, err := c.db.Prepare("select value from config where key=?")
+	stmt, err := c.PrepareQuery("select value from config where key=?")
 	if err != nil {
 		return nil, err
 	}
@@ -327,7 +308,7 @@ func (c *ConfigDB) GetCreationDate() (time.Time, error) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
 	var t time.Time
-	stmt, err := c.db.Prepare("select value from config where key=?")
+	stmt, err := c.PrepareQuery("select value from config where key=?")
 	if err != nil {
 		return t, err
 	}

--- a/repo/db/followers.go
+++ b/repo/db/followers.go
@@ -2,6 +2,7 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
 	"strconv"
 	"sync"
 
@@ -19,21 +20,15 @@ func NewFollowerStore(db *sql.DB, lock *sync.Mutex) repo.FollowerStore {
 func (f *FollowerDB) Put(follower string, proof []byte) error {
 	f.lock.Lock()
 	defer f.lock.Unlock()
-	tx, _ := f.db.Begin()
-	stmt, _ := tx.Prepare("insert into followers(peerID, proof) values(?,?)")
+	stmt, err := f.PrepareQuery("insert into followers(peerID, proof) values(?,?)")
+	if err != nil {
+		return fmt.Errorf("prepare followers sql: %s", err.Error())
+	}
 
 	defer stmt.Close()
-	_, err := stmt.Exec(follower, proof)
+	_, err = stmt.Exec(follower, proof)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("update followers: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/following.go
+++ b/repo/db/following.go
@@ -2,6 +2,7 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
 	"strconv"
 	"sync"
 
@@ -19,20 +20,14 @@ func NewFollowingStore(db *sql.DB, lock *sync.Mutex) repo.FollowingStore {
 func (f *FollowingDB) Put(follower string) error {
 	f.lock.Lock()
 	defer f.lock.Unlock()
-	tx, _ := f.db.Begin()
-	stmt, _ := tx.Prepare("insert into following(peerID) values(?)")
-	defer stmt.Close()
-	_, err := stmt.Exec(follower)
+	stmt, err := f.PrepareQuery("insert into following(peerID) values(?)")
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("prepare following sql: %s", err.Error())
 	}
-	err = tx.Commit()
+	defer stmt.Close()
+	_, err = stmt.Exec(follower)
 	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit following: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/inventory.go
+++ b/repo/db/inventory.go
@@ -5,6 +5,7 @@ import (
 	"database/sql"
 	"encoding/hex"
 	"errors"
+	"fmt"
 	"math/big"
 	"strconv"
 	"sync"
@@ -26,23 +27,14 @@ func (i *InventoryDB) Put(slug string, variantIndex int, count *big.Int) error {
 
 	id := sha256.Sum256([]byte(slug + strconv.Itoa(variantIndex)))
 
-	tx, _ := i.db.Begin()
-	stmt, err := tx.Prepare("insert or replace into inventory(invID, slug, variantIndex, count) values(?,?,?,?)")
+	stmt, err := i.PrepareQuery("insert or replace into inventory(invID, slug, variantIndex, count) values(?,?,?,?)")
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare inventory sql: %s", err.Error())
 	}
 	defer stmt.Close()
 	_, err = stmt.Exec(hex.EncodeToString(id[:]), slug, variantIndex, count.String())
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("update inventory: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/keys.go
+++ b/repo/db/keys.go
@@ -4,6 +4,7 @@ import (
 	"database/sql"
 	"encoding/hex"
 	"errors"
+	"fmt"
 	"strconv"
 	"sync"
 
@@ -24,26 +25,15 @@ func NewKeyStore(db *sql.DB, lock *sync.Mutex, coinType wallet.CoinType) repo.Ke
 func (k *KeysDB) Put(scriptAddress []byte, keyPath wallet.KeyPath) error {
 	k.lock.Lock()
 	defer k.lock.Unlock()
-	tx, err := k.db.Begin()
+	stmt, err := k.PrepareQuery("insert into keys(coin, scriptAddress, purpose, keyIndex, used) values(?,?,?,?,?)")
 	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("insert into keys(coin, scriptAddress, purpose, keyIndex, used) values(?,?,?,?,?)")
-	if err != nil {
-		return err
+		return fmt.Errorf("prepare key sql: %s", err.Error())
 	}
 	defer stmt.Close()
+
 	_, err = stmt.Exec(k.coinType.CurrencyCode(), hex.EncodeToString(scriptAddress), int(keyPath.Purpose), keyPath.Index, 0)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit key: %s", err.Error())
 	}
 	return nil
 }
@@ -51,26 +41,14 @@ func (k *KeysDB) Put(scriptAddress []byte, keyPath wallet.KeyPath) error {
 func (k *KeysDB) ImportKey(scriptAddress []byte, key *btcec.PrivateKey) error {
 	k.lock.Lock()
 	defer k.lock.Unlock()
-	tx, err := k.db.Begin()
+	stmt, err := k.PrepareQuery("insert into keys(coin, scriptAddress, purpose, used, key) values(?,?,?,?,?)")
 	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("insert into keys(coin, scriptAddress, purpose, used, key) values(?,?,?,?,?)")
-	if err != nil {
-		return err
+		return fmt.Errorf("prepare key sql: %s", err.Error())
 	}
 	defer stmt.Close()
 	_, err = stmt.Exec(k.coinType.CurrencyCode(), hex.EncodeToString(scriptAddress), -1, 0, hex.EncodeToString(key.Serialize()))
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit key: %s", err.Error())
 	}
 	return nil
 }
@@ -78,27 +56,15 @@ func (k *KeysDB) ImportKey(scriptAddress []byte, key *btcec.PrivateKey) error {
 func (k *KeysDB) MarkKeyAsUsed(scriptAddress []byte) error {
 	k.lock.Lock()
 	defer k.lock.Unlock()
-	tx, err := k.db.Begin()
+	stmt, err := k.PrepareQuery("update keys set used=1 where scriptAddress=? and coin=?")
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare key sql: %s", err.Error())
 	}
-	stmt, err := tx.Prepare("update keys set used=1 where scriptAddress=? and coin=?")
-	if err != nil {
-		return err
-	}
-
 	defer stmt.Close()
+
 	_, err = stmt.Exec(hex.EncodeToString(scriptAddress), k.coinType.CurrencyCode())
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit key update: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/messages.go
+++ b/repo/db/messages.go
@@ -25,19 +25,15 @@ func (o *MessagesDB) Put(messageID, orderID string, mType pb.Message_MessageType
 	o.lock.Lock()
 	defer o.lock.Unlock()
 
-	tx, err := o.db.Begin()
-	if err != nil {
-		return err
-	}
 	stm := `insert or replace into messages(messageID, orderID, message_type, message, peerID, created_at) values(?,?,?,?,?,?)`
-	stmt, err := tx.Prepare(stm)
+	stmt, err := o.PrepareQuery(stm)
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare message sql: %s", err.Error())
 	}
 
 	msg0, err := msg.MarshalJSON()
 	if err != nil {
-		log.Errorf("err marshalling json: %v", err)
+		return fmt.Errorf("marshal message: %s", err.Error())
 	}
 
 	defer stmt.Close()
@@ -50,14 +46,10 @@ func (o *MessagesDB) Put(messageID, orderID string, mType pb.Message_MessageType
 		int(time.Now().Unix()),
 	)
 	if err != nil {
-		rErr := tx.Rollback()
-		if rErr != nil {
-			return fmt.Errorf("message put fail: %s and rollback failed: %s", err.Error(), rErr.Error())
-		}
-		return err
+		return fmt.Errorf("commit message: %s", err.Error())
 	}
 
-	return tx.Commit()
+	return nil
 }
 
 // GetByOrderIDType returns the message for the specified order and message type

--- a/repo/db/moderatedstores.go
+++ b/repo/db/moderatedstores.go
@@ -2,6 +2,7 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
 	"strconv"
 	"sync"
 
@@ -19,21 +20,15 @@ func NewModeratedStore(db *sql.DB, lock *sync.Mutex) repo.ModeratedStore {
 func (m *ModeratedDB) Put(peerId string) error {
 	m.lock.Lock()
 	defer m.lock.Unlock()
-	tx, _ := m.db.Begin()
-	stmt, _ := tx.Prepare("insert into moderatedstores(peerID) values(?)")
-
-	defer stmt.Close()
-	_, err := stmt.Exec(peerId)
+	stmt, err := m.PrepareQuery("insert into moderatedstores(peerID) values(?)")
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("prepare moderated store sql: %s", err.Error())
 	}
-	err = tx.Commit()
+	defer stmt.Close()
+
+	_, err = stmt.Exec(peerId)
 	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit moderated store: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/notifications.go
+++ b/repo/db/notifications.go
@@ -134,24 +134,15 @@ func (n *NotficationsDB) GetAll(offsetId string, limit int, typeFilter []string)
 func (n *NotficationsDB) MarkAsRead(notifID string) error {
 	n.lock.Lock()
 	defer n.lock.Unlock()
-	tx, err := n.db.Begin()
+	stmt, err := n.PrepareQuery("update notifications set read=1 where notifID=?")
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare notification sql: %s", err.Error())
 	}
-	stmt, _ := tx.Prepare("update notifications set read=1 where notifID=?")
-
 	defer stmt.Close()
+
 	_, err = stmt.Exec(notifID)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit notification as read: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/offinemessages.go
+++ b/repo/db/offinemessages.go
@@ -2,6 +2,7 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
 	"sync"
 	"time"
 
@@ -19,26 +20,15 @@ func NewOfflineMessageStore(db *sql.DB, lock *sync.Mutex) repo.OfflineMessageSto
 func (o *OfflineMessagesDB) Put(url string) error {
 	o.lock.Lock()
 	defer o.lock.Unlock()
-	tx, err := o.db.Begin()
+	stmt, err := o.PrepareQuery("insert into offlinemessages(url, timestamp) values(?,?)")
 	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("insert into offlinemessages(url, timestamp) values(?,?)")
-	if err != nil {
-		return err
+		return fmt.Errorf("prepare offline message sql: %s", err.Error())
 	}
 	defer stmt.Close()
+
 	_, err = stmt.Exec(url, int(time.Now().Unix()))
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit offline message: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/pointers.go
+++ b/repo/db/pointers.go
@@ -2,15 +2,15 @@ package db
 
 import (
 	"database/sql"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
 
 	ma "gx/ipfs/QmTZBfrPJmjWsCvHEtX5FE6KimVJhsJg5sBbqEFYf4UZtL/go-multiaddr"
 	cid "gx/ipfs/QmTbxNB1NwDesLmKTscr4udL2tVP7MaxvXnD1D9yX7g3PN/go-cid"
 	peer "gx/ipfs/QmYVXrKrKHDC9FobgmcmshCDyWwdrfwfanNQN4oxJ9Fk3h/go-libp2p-peer"
 	ps "gx/ipfs/QmaCTz9RkrU13bm9kMB54f7atgqM4qkjDZpRwRoJiWXEqs/go-libp2p-peerstore"
-
-	"strconv"
-	"sync"
-	"time"
 
 	"github.com/OpenBazaar/openbazaar-go/ipfs"
 	"github.com/OpenBazaar/openbazaar-go/repo"
@@ -27,30 +27,20 @@ func NewPointerStore(db *sql.DB, lock *sync.Mutex) repo.PointerStore {
 func (p *PointersDB) Put(pointer ipfs.Pointer) error {
 	p.lock.Lock()
 	defer p.lock.Unlock()
-	tx, err := p.db.Begin()
+
+	stmt, err := p.PrepareQuery("insert into pointers(pointerID, key, address, cancelID, purpose, timestamp) values(?,?,?,?,?,?)")
 	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("insert into pointers(pointerID, key, address, cancelID, purpose, timestamp) values(?,?,?,?,?,?)")
-	if err != nil {
-		return err
+		return fmt.Errorf("prepare pointer sql: %s", err.Error())
 	}
 	defer stmt.Close()
+
 	var cancelID string
 	if pointer.CancelID != nil {
 		cancelID = pointer.CancelID.Pretty()
 	}
 	_, err = stmt.Exec(pointer.Value.ID.Pretty(), pointer.Cid.String(), pointer.Value.Addrs[0].String(), cancelID, pointer.Purpose, int(time.Now().Unix()))
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit pointer: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/purchases.go
+++ b/repo/db/purchases.go
@@ -330,9 +330,11 @@ func (p *PurchasesDB) GetByOrderId(orderId string) (*pb.RicardianContract, pb.Or
 		return nil, pb.OrderState(0), false, nil, false, nil, fmt.Errorf("validating payment coin: %s", err.Error())
 	}
 	var records []*wallet.TransactionRecord
-	err = json.Unmarshal(serializedTransactions, &records)
-	if err != nil {
-		return nil, pb.OrderState(0), false, nil, false, nil, fmt.Errorf("unmarshal purchase transactions: %s", err.Error())
+	if len(serializedTransactions) > 0 {
+		err = json.Unmarshal(serializedTransactions, &records)
+		if err != nil {
+			return nil, pb.OrderState(0), false, nil, false, nil, fmt.Errorf("unmarshal purchase transactions: %s", err.Error())
+		}
 	}
 	cc := def.CurrencyCode()
 	return rc, pb.OrderState(stateInt), funded, records, read, &cc, nil

--- a/repo/db/purchases.go
+++ b/repo/db/purchases.go
@@ -41,15 +41,12 @@ func (p *PurchasesDB) Put(orderID string, contract pb.RicardianContract, state p
 		return err
 	}
 
-	tx, err := p.db.Begin()
-	if err != nil {
-		return err
-	}
 	stm := `insert or replace into purchases(orderID, contract, state, read, timestamp, total, thumbnail, vendorID, vendorHandle, title, shippingName, shippingAddress, paymentAddr, paymentCoin, coinType, funded, transactions, disputedAt) values(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,(select funded from purchases where orderID="` + orderID + `"),(select transactions from purchases where orderID="` + orderID + `"),?)`
-	stmt, err := tx.Prepare(stm)
+	stmt, err := p.db.Prepare(stm)
 	if err != nil {
 		return err
 	}
+	defer stmt.Close()
 	var (
 		paymentAddr, shippingName, shippingAddress string
 		disputedAt                                 int
@@ -70,8 +67,6 @@ func (p *PurchasesDB) Put(orderID string, contract pb.RicardianContract, state p
 	if dispute != nil {
 		disputedAt = int(dispute.Timestamp.Seconds)
 	}
-
-	defer stmt.Close()
 	_, err = stmt.Exec(
 		orderID,
 		out,
@@ -91,13 +86,9 @@ func (p *PurchasesDB) Put(orderID string, contract pb.RicardianContract, state p
 		disputedAt,
 	)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("commit purchase: %s", err.Error())
 	}
-	return tx.Commit()
+	return nil
 }
 
 func (p *PurchasesDB) MarkAsRead(orderID string) error {
@@ -341,7 +332,7 @@ func (p *PurchasesDB) GetByOrderId(orderId string) (*pb.RicardianContract, pb.Or
 	var records []*wallet.TransactionRecord
 	err = json.Unmarshal(serializedTransactions, &records)
 	if err != nil {
-		log.Error(err)
+		return nil, pb.OrderState(0), false, nil, false, nil, fmt.Errorf("unmarshal purchase transactions: %s", err.Error())
 	}
 	cc := def.CurrencyCode()
 	return rc, pb.OrderState(stateInt), funded, records, read, &cc, nil
@@ -354,7 +345,8 @@ func (p *PurchasesDB) Count() int {
 	var count int
 	err := row.Scan(&count)
 	if err != nil {
-		log.Error(err)
+		log.Errorf("failed scanning purchase count: %s", err.Error())
+		return 0
 	}
 	return count
 }
@@ -463,17 +455,23 @@ func (p *PurchasesDB) UpdatePurchasesLastDisputeTimeoutNotifiedAt(purchases []*r
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	tx, err := p.db.Begin()
+	tx, err := p.BeginTransaction()
 	if err != nil {
 		return fmt.Errorf("begin update purchase transaction: %s", err.Error())
 	}
 	for _, p := range purchases {
 		_, err = tx.Exec("update purchases set lastDisputeTimeoutNotifiedAt = ? where orderID = ?", int(p.LastDisputeTimeoutNotifiedAt.Unix()), p.OrderID)
 		if err != nil {
+			if rErr := tx.Rollback(); rErr != nil {
+				return fmt.Errorf("update purchase: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+			}
 			return fmt.Errorf("update purchase: %s", err.Error())
 		}
 	}
 	if err = tx.Commit(); err != nil {
+		if rErr := tx.Rollback(); rErr != nil {
+			return fmt.Errorf("commit purchase: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+		}
 		return fmt.Errorf("commit update purchase transaction: %s", err.Error())
 	}
 
@@ -488,17 +486,23 @@ func (p *PurchasesDB) UpdatePurchasesLastDisputeExpiryNotifiedAt(purchases []*re
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
-	tx, err := p.db.Begin()
+	tx, err := p.BeginTransaction()
 	if err != nil {
 		return fmt.Errorf("begin update purchase transaction: %s", err.Error())
 	}
 	for _, p := range purchases {
 		_, err = tx.Exec("update purchases set lastDisputeExpiryNotifiedAt = ? where orderID = ?", int(p.LastDisputeExpiryNotifiedAt.Unix()), p.OrderID)
 		if err != nil {
+			if rErr := tx.Rollback(); rErr != nil {
+				return fmt.Errorf("update purchase error: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+			}
 			return fmt.Errorf("update purchase: %s", err.Error())
 		}
 	}
 	if err = tx.Commit(); err != nil {
+		if rErr := tx.Rollback(); rErr != nil {
+			return fmt.Errorf("commit purchase error: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+		}
 		return fmt.Errorf("commit update purchase transaction: %s", err.Error())
 	}
 

--- a/repo/db/sales.go
+++ b/repo/db/sales.go
@@ -46,15 +46,12 @@ func (s *SalesDB) Put(orderID string, contract pb.RicardianContract, state pb.Or
 		return err
 	}
 
-	tx, err := s.db.Begin()
-	if err != nil {
-		return err
-	}
 	stm := `insert or replace into sales(orderID, contract, state, read, timestamp, total, thumbnail, buyerID, buyerHandle, title, shippingName, shippingAddress, paymentAddr, paymentCoin, coinType, funded, transactions) values(?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,(select funded from sales where orderID="` + orderID + `"),(select transactions from sales where orderID="` + orderID + `"))`
-	stmt, err := tx.Prepare(stm)
+	stmt, err := s.db.Prepare(stm)
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare sale sql: %s", err.Error())
 	}
+	defer stmt.Close()
 
 	handle := contract.BuyerOrder.BuyerID.Handle
 	shippingName := ""
@@ -70,7 +67,6 @@ func (s *SalesDB) Put(orderID string, contract pb.RicardianContract, state pb.Or
 		address = contract.VendorOrderConfirmation.PaymentAddress
 	}
 
-	defer stmt.Close()
 	_, err = stmt.Exec(
 		orderID,
 		out,
@@ -89,14 +85,10 @@ func (s *SalesDB) Put(orderID string, contract pb.RicardianContract, state pb.Or
 		CoinTypeForContract(&contract),
 	)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("commit sale: %s", err.Error())
 	}
 
-	return tx.Commit()
+	return nil
 }
 
 func (s *SalesDB) MarkAsRead(orderID string) error {
@@ -420,18 +412,24 @@ func (s *SalesDB) UpdateSalesLastDisputeTimeoutNotifiedAt(sales []*repo.SaleReco
 	s.lock.Lock()
 	defer s.lock.Unlock()
 
-	tx, err := s.db.Begin()
+	tx, err := s.BeginTransaction()
 	if err != nil {
 		return fmt.Errorf("begin update sale transaction: %s", err.Error())
 	}
 	for _, sale := range sales {
 		_, err = tx.Exec("update sales set lastDisputeTimeoutNotifiedAt = ? where orderID = ?", int(sale.LastDisputeTimeoutNotifiedAt.Unix()), sale.OrderID)
 		if err != nil {
+			if rErr := tx.Rollback(); rErr != nil {
+				return fmt.Errorf("update sale: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+			}
 			return fmt.Errorf("update sale: %s", err.Error())
 		}
 	}
 	if err = tx.Commit(); err != nil {
-		return fmt.Errorf("commit update sale transaction: %s", err.Error())
+		if rErr := tx.Rollback(); rErr != nil {
+			return fmt.Errorf("commit sale: (%s) w rollback error: (%s)", err.Error(), rErr.Error())
+		}
+		return fmt.Errorf("commit sale: %s", err.Error())
 	}
 
 	return nil

--- a/repo/db/settings.go
+++ b/repo/db/settings.go
@@ -23,29 +23,20 @@ func NewConfigurationStore(db *sql.DB, lock *sync.Mutex) repo.ConfigurationStore
 func (s *SettingsDB) Put(settings repo.SettingsData) error {
 	s.lock.Lock()
 	defer s.lock.Unlock()
-	tx, err := s.db.Begin()
-	if err != nil {
-		return err
-	}
+
 	b, err := json.MarshalIndent(&settings, "", "    ")
 	if err != nil {
-		return err
+		return fmt.Errorf("marshal settings: %s", err.Error())
 	}
-	stmt, err := tx.Prepare("insert or replace into config(key, value) values(?,?)")
+	stmt, err := s.PrepareQuery("insert or replace into config(key, value) values(?,?)")
 	if err != nil {
-		return err
+		return fmt.Errorf("prepare settings sql: %s", err.Error())
 	}
 	defer stmt.Close()
 
 	_, err = stmt.Exec("settings", string(b))
 	if err != nil {
-		if errRollback := tx.Rollback(); errRollback != nil {
-			return fmt.Errorf("rollback: %s\n also: %s", errRollback.Error(), err.Error())
-		}
-		return err
-	}
-	if err := tx.Commit(); err != nil {
-		return err
+		return fmt.Errorf("commit settings: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/stxo.go
+++ b/repo/db/stxo.go
@@ -3,6 +3,7 @@ package db
 import (
 	"database/sql"
 	"encoding/hex"
+	"fmt"
 	"strconv"
 	"strings"
 	"sync"
@@ -25,16 +26,12 @@ func NewSpentTransactionStore(db *sql.DB, lock *sync.Mutex, coinType wallet.Coin
 func (s *StxoDB) Put(stxo wallet.Stxo) error {
 	s.lock.Lock()
 	defer s.lock.Unlock()
-	tx, _ := s.db.Begin()
-	stmt, err := tx.Prepare("insert or replace into stxos(coin, outpoint, value, height, scriptPubKey, watchOnly, spendHeight, spendTxid) values(?,?,?,?,?,?,?,?)")
+	stmt, err := s.PrepareQuery("insert or replace into stxos(coin, outpoint, value, height, scriptPubKey, watchOnly, spendHeight, spendTxid) values(?,?,?,?,?,?,?,?)")
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("prepare stxo sql: %s", err.Error())
 	}
 	defer stmt.Close()
+
 	watchOnly := 0
 	if stxo.Utxo.WatchOnly {
 		watchOnly = 1
@@ -42,15 +39,7 @@ func (s *StxoDB) Put(stxo wallet.Stxo) error {
 	outpoint := stxo.Utxo.Op.Hash.String() + ":" + strconv.Itoa(int(stxo.Utxo.Op.Index))
 	_, err = stmt.Exec(s.coinType.CurrencyCode(), outpoint, stxo.Utxo.Value, int(stxo.Utxo.AtHeight), hex.EncodeToString(stxo.Utxo.ScriptPubkey), watchOnly, int(stxo.SpendHeight), stxo.SpendTxid.String())
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit stxo: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/txmetadata.go
+++ b/repo/db/txmetadata.go
@@ -18,13 +18,9 @@ func NewTransactionMetadataStore(db *sql.DB, lock *sync.Mutex) repo.TransactionM
 func (t *TxMetadataDB) Put(m repo.Metadata) error {
 	t.lock.Lock()
 	defer t.lock.Unlock()
-	tx, _ := t.db.Begin()
-	stmt, err := tx.Prepare("insert or replace into txmetadata(txid, address, memo, orderID, thumbnail, canBumpFee) values(?,?,?,?,?,?)")
+	stmt, err := t.db.Prepare("insert or replace into txmetadata(txid, address, memo, orderID, thumbnail, canBumpFee) values(?,?,?,?,?,?)")
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
+		log.Errorf("prepring txmetadata sql for order (%s): %s", m.OrderId, err.Error())
 		return err
 	}
 	defer stmt.Close()
@@ -34,15 +30,8 @@ func (t *TxMetadataDB) Put(m repo.Metadata) error {
 	}
 	_, err = stmt.Exec(m.Txid, m.Address, m.Memo, m.OrderId, m.Thumbnail, bumpable)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
+		log.Errorf("putting txmetadata for order (%s): %s", m.OrderId, err.Error())
 		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
 	}
 	return nil
 }

--- a/repo/db/txns.go
+++ b/repo/db/txns.go
@@ -23,34 +23,20 @@ func NewTransactionStore(db *sql.DB, lock *sync.Mutex, coinType wallet.CoinType)
 func (t *TxnsDB) Put(raw []byte, txid, value string, height int, timestamp time.Time, watchOnly bool) error {
 	t.lock.Lock()
 	defer t.lock.Unlock()
-	tx, err := t.db.Begin()
+
+	stmt, err := t.PrepareQuery("insert or replace into txns(coin, txid, value, height, timestamp, watchOnly, tx) values(?,?,?,?,?,?,?)")
 	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("insert or replace into txns(coin, txid, value, height, timestamp, watchOnly, tx) values(?,?,?,?,?,?,?)")
-	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("prepare txn sql: %s", err.Error())
 	}
 	defer stmt.Close()
+
 	watchOnlyInt := 0
 	if watchOnly {
 		watchOnlyInt = 1
 	}
 	_, err = stmt.Exec(t.coinType.CurrencyCode(), txid, value, height, int(timestamp.Unix()), watchOnlyInt, raw)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("update txn: %s", err.Error())
 	}
 	return nil
 }
@@ -144,27 +130,15 @@ func (t *TxnsDB) Delete(txid *chainhash.Hash) error {
 func (t *TxnsDB) UpdateHeight(txid chainhash.Hash, height int, timestamp time.Time) error {
 	t.lock.Lock()
 	defer t.lock.Unlock()
-	tx, err := t.db.Begin()
+
+	stmt, err := t.PrepareQuery("update txns set height=?, timestamp=? where txid=? and coin=?")
 	if err != nil {
-		return err
-	}
-	stmt, err := tx.Prepare("update txns set height=?, timestamp=? where txid=? and coin=?")
-	if err != nil {
-		if rErr := tx.Rollback(); rErr != nil {
-			return fmt.Errorf("%s (db rollback: %s)", err.Error(), rErr.Error())
-		}
-		return err
+		return fmt.Errorf("prepare txn sql: %s", err.Error())
 	}
 	defer stmt.Close()
 	_, err = stmt.Exec(height, int(timestamp.Unix()), txid.String(), t.coinType.CurrencyCode())
 	if err != nil {
-		if rErr := tx.Rollback(); rErr != nil {
-			return fmt.Errorf("%s (db rollback: %s)", err.Error(), rErr.Error())
-		}
-		return err
-	}
-	if err := tx.Commit(); err != nil {
-		return err
+		return fmt.Errorf("update txns: %s", err.Error())
 	}
 	return nil
 }

--- a/repo/db/utxos.go
+++ b/repo/db/utxos.go
@@ -3,6 +3,7 @@ package db
 import (
 	"database/sql"
 	"encoding/hex"
+	"fmt"
 	"strconv"
 	"strings"
 	"sync"
@@ -25,16 +26,12 @@ func NewUnspentTransactionStore(db *sql.DB, lock *sync.Mutex, coinType wallet.Co
 func (u *UtxoDB) Put(utxo wallet.Utxo) error {
 	u.lock.Lock()
 	defer u.lock.Unlock()
-	tx, _ := u.db.Begin()
-	stmt, err := tx.Prepare("insert or replace into utxos(coin, outpoint, value, height, scriptPubKey, watchOnly) values(?,?,?,?,?,?)")
+	stmt, err := u.PrepareQuery("insert or replace into utxos(coin, outpoint, value, height, scriptPubKey, watchOnly) values(?,?,?,?,?,?)")
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("prepare utxo sql: %s", err.Error())
 	}
 	defer stmt.Close()
+
 	watchOnlyInt := 0
 	if utxo.WatchOnly {
 		watchOnlyInt = 1
@@ -42,15 +39,7 @@ func (u *UtxoDB) Put(utxo wallet.Utxo) error {
 	outpoint := utxo.Op.Hash.String() + ":" + strconv.Itoa(int(utxo.Op.Index))
 	_, err = stmt.Exec(u.coinType.CurrencyCode(), outpoint, utxo.Value, int(utxo.AtHeight), hex.EncodeToString(utxo.ScriptPubkey), watchOnlyInt)
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit utxo: %s", err.Error())
 	}
 	return nil
 }
@@ -62,7 +51,7 @@ func (u *UtxoDB) GetAll() ([]wallet.Utxo, error) {
 	stm := "select outpoint, value, height, scriptPubKey, watchOnly from utxos where coin=?"
 	rows, err := u.db.Query(stm, u.coinType.CurrencyCode())
 	if err != nil {
-		return ret, err
+		return ret, fmt.Errorf("get utxos: %s", err.Error())
 	}
 	defer rows.Close()
 	for rows.Next() {
@@ -72,19 +61,23 @@ func (u *UtxoDB) GetAll() ([]wallet.Utxo, error) {
 		var scriptPubKey string
 		var watchOnlyInt int
 		if err := rows.Scan(&outpoint, &value, &height, &scriptPubKey, &watchOnlyInt); err != nil {
+			log.Errorf("scanning utxo record: %s", err.Error())
 			continue
 		}
 		s := strings.Split(outpoint, ":")
 		shaHash, err := chainhash.NewHashFromStr(s[0])
 		if err != nil {
+			log.Errorf("getting outpoint hash: %s", err.Error())
 			continue
 		}
 		index, err := strconv.Atoi(s[1])
 		if err != nil {
+			log.Errorf("getting output index: %s", err.Error())
 			continue
 		}
 		scriptBytes, err := hex.DecodeString(scriptPubKey)
 		if err != nil {
+			log.Errorf("decode redeem script pubkey: %s", err.Error())
 			continue
 		}
 		watchOnly := false

--- a/repo/db/watched_scripts.go
+++ b/repo/db/watched_scripts.go
@@ -3,6 +3,7 @@ package db
 import (
 	"database/sql"
 	"encoding/hex"
+	"fmt"
 	"sync"
 
 	"github.com/OpenBazaar/openbazaar-go/repo"
@@ -21,27 +22,14 @@ func NewWatchedScriptStore(db *sql.DB, lock *sync.Mutex, coinType wallet.CoinTyp
 func (w *WatchedScriptsDB) Put(scriptPubKey []byte) error {
 	w.lock.Lock()
 	defer w.lock.Unlock()
-	tx, _ := w.db.Begin()
-	stmt, err := tx.Prepare("insert or replace into watchedscripts(coin, scriptPubKey) values(?,?)")
+	stmt, err := w.PrepareQuery("insert or replace into watchedscripts(coin, scriptPubKey) values(?,?)")
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
+		return fmt.Errorf("prepare watch script sql: %s", err.Error())
 	}
 	defer stmt.Close()
 	_, err = stmt.Exec(w.coinType.CurrencyCode(), hex.EncodeToString(scriptPubKey))
 	if err != nil {
-		err0 := tx.Rollback()
-		if err0 != nil {
-			log.Error(err0)
-		}
-		return err
-	}
-	err = tx.Commit()
-	if err != nil {
-		log.Error(err)
+		return fmt.Errorf("commit watch script: %s", err.Error())
 	}
 	return nil
 }
@@ -53,16 +41,18 @@ func (w *WatchedScriptsDB) GetAll() ([][]byte, error) {
 	stm := "select scriptPubKey from watchedscripts where coin=?"
 	rows, err := w.db.Query(stm, w.coinType.CurrencyCode())
 	if err != nil {
-		return ret, err
+		return nil, err
 	}
 	defer rows.Close()
 	for rows.Next() {
 		var scriptHex string
 		if err := rows.Scan(&scriptHex); err != nil {
+			log.Errorf("scan watch script key: %s", err.Error())
 			continue
 		}
 		scriptPubKey, err := hex.DecodeString(scriptHex)
 		if err != nil {
+			log.Errorf("decode watch script key: %s", err.Error())
 			continue
 		}
 		ret = append(ret, scriptPubKey)
@@ -75,7 +65,7 @@ func (w *WatchedScriptsDB) Delete(scriptPubKey []byte) error {
 	defer w.lock.Unlock()
 	_, err := w.db.Exec("delete from watchedscripts where scriptPubKey=? and coin=?", hex.EncodeToString(scriptPubKey), w.coinType.CurrencyCode())
 	if err != nil {
-		return err
+		return fmt.Errorf("delete watch script key: %s", err.Error())
 	}
 	return nil
 }

